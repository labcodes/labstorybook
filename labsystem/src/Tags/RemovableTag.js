import React from "react";
import PropTypes from "prop-types";
import { isEmpty } from "lodash";
import AbstractTag from "./AbstractTag";
import Icon from "../Icon";

export default class RemovableTag extends React.Component {
  static propTypes = {
    /** Text that will be rendered inside the Tag. */
    text: PropTypes.string.isRequired,
    /** Source of the thumb to be rendered. Won't render a thumb if not passed to the component. */
    thumbSrc: PropTypes.string,
    /** Type of the icon to be rendered. Won't render an icon if not passed to the component. */
    icon: PropTypes.string,
    /** Color of the rendered Tag. */
    color: PropTypes.string,
    /** Skin of the the rendered Tag. */
    skin: PropTypes.string,
    /** Defines if the Tag has an outline style. */
    isOutline: PropTypes.bool,
    /** Disables the Tag component, including functionality and style. */
    disabled: PropTypes.bool,
    /** Callback to be executed when the Tag is clicked. */
    onClick: PropTypes.func,
  };

  static defaultProps = {
    thumbSrc: "",
    icon: "",
    color: "",
    skin: "pale",
    isOutline: false,
    disabled: false,
    onClick: () => {},
  };

  constructor(props) {
    super(props);
    this.checkThumbAndIcon();
  }

  componentDidUpdate() {
    this.checkThumbAndIcon();
  }

  removeIcon = () => (
    <span className="lab-tag__remove-icon-wrapper">
      <Icon
        type="remove"
        color="black-75"
        size="petit"
        className="lab-tag--remove-icon"
      />
    </span>
  );

  thumb = () => {
    const { thumbSrc } = this.props;
    return thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt="" />
    ) : undefined;
  };

  icon = () => {
    const { icon } = this.props;
    return icon ? (
      <Icon
        type={icon}
        color="black-75"
        size="petit"
        className="lab-tag--left-icon"
      />
    ) : undefined;
  };

  checkThumbAndIcon() {
    const errorMessage =
      "`RemovableTag` can't be initialized with both `thumb` and `icon` props.";
    const { thumbSrc, icon } = this.props;
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw Error(errorMessage);
    }
  }

  render() {
    const {
      text,
      thumbSrc,
      icon,
      color,
      skin,
      isOutline,
      disabled,
      onClick,
    } = this.props;
    return (
      <AbstractTag
        className={`lab-tag--removable${`${
          icon ? ` lab-tag--has-left-icon` : ""
        }${thumbSrc ? ` lab-tag--has-thumb` : ""}`}`}
        text={text}
        thumbSrc={thumbSrc}
        icon={icon}
        color={color}
        skin={skin}
        isOutline={isOutline}
        disabled={disabled}
        onClick={onClick}
        renderPrefix={this.thumb() || this.icon()}
        renderSuffix={this.removeIcon()}
      />
    );
  }
}
