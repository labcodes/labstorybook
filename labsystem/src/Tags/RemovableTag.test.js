import React from "react";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

import RemovableTag from "./RemovableTag";

describe("RemovableTag", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<RemovableTag text="Test RemovableTag" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as disabled", async () => {
    const renderedComponent = renderer
      .create(
        <RemovableTag text="Test render a disabled RemovableTag" isDisabled />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <RemovableTag text="Test mount disabled RemovableTag" isDisabled />
    );
    expect(mountedComponent.find("AbstractTag").prop("isDisabled")).toEqual(
      true
    );
  });

  it("renders as outline", async () => {
    const renderedComponent = renderer
      .create(
        <RemovableTag text="Test render an outline RemovableTag" isOutline />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <RemovableTag text="Test mount outline RemovableTag" isOutline />
    );
    expect(mountedComponent.find("AbstractTag").prop("isOutline")).toEqual(
      true
    );
  });

  it("renders with a green color", async () => {
    const renderedComponent = renderer
      .create(
        <RemovableTag text="Test render green RemovableTag" color="green" />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <RemovableTag text="Test mount green RemovableTag" color="green" />
    );
    expect(mountedComponent.find("AbstractTag").prop("color")).toEqual("green");
  });

  it("calls prop.onClick when clicked", async () => {
    const mockOnClick = jest.fn();
    const shallowedRemovableTag = shallow(
      <RemovableTag text="Test click on RemovableTag" onClick={mockOnClick} />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    shallowedRemovableTag.simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  const thumbSrcSample =
    "https://avatars3.githubusercontent.com/u/1887591?s=400&u=ba45b6433752099210bf180b4448fb82e015c3a8&v=4";
  it("does not render if passing both `thumb` and `icon`", async () => {
    expect(() => {
      shallow(
        <RemovableTag
          text="Test to not render RemovableTag with thumb and icon"
          icon="calendar"
          thumbSrc={thumbSrcSample}
        />
      );
    }).toThrow(
      "`RemovableTag` can't be initialized with both `thumb` and `icon` props."
    );
  });
});
