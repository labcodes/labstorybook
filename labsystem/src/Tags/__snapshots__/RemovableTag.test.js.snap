// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemovableTag renders as disabled 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--disabled lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="presentation"
>
  Test render a disabled RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--remove lab-icon--black75 lab-icon--petit lab-icon--remove-icon"
    />
  </span>
</span>
`;

exports[`RemovableTag renders as outline 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--outline lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="presentation"
>
  Test render an outline RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--remove lab-icon--black75 lab-icon--petit lab-icon--remove-icon"
    />
  </span>
</span>
`;

exports[`RemovableTag renders with a green color 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--green-pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="presentation"
>
  Test render green RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--remove lab-icon--black75 lab-icon--petit lab-icon--remove-icon"
    />
  </span>
</span>
`;

exports[`RemovableTag renders with base props 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="presentation"
>
  Test RemovableTag
  <span
    className="lab-tag__remove-icon-wrapper"
  >
    <span
      className="lab-icon lab-icon--remove lab-icon--black75 lab-icon--petit lab-icon--remove-icon"
    />
  </span>
</span>
`;
